Tests/MongoSwiftTests/DocumentTests.swift

```diff
@@ -305,126 +305,244 @@ final class DocumentTests: MongoSwiftTestCase {
-                if let dB = validCase["degenerate_bson"] as? String {
-                    guard let dBData = Data(hexString: dB) else {
-                        XCTFail("Unable to interpret degenerate_bson as Data")
-                        return
+                        // native_to_bson( json_to_native(dEJ) ) = cB (unless lossy)
+                        if !lossy {
+                            expect(try Document(fromJSON: dEJ).rawBSON).to(equal(cBData))
+                        }
                     }
 
-                    // bson_to_canonical_extended_json(dB) = cEJ
-                    expect(try Document(fromBSON: dBData).canonicalExtendedJSON).to(cleanEqual(cEJ))
+                    // for rEJ input (if it exists):
+                    if let rEJ = test.relaxedExtJSON {
+                        // native_to_relaxed_extended_json( json_to_native(rEJ) ) = rEJ
+                        expect(try Document(fromJSON: rEJ).extendedJSON).to(cleanEqual(rEJ))
+                    }
+                }
+            }
 
-                    // bson_to_relaxed_extended_json(dB) = rEJ (if rEJ exists)
-                    if let rEJ = validCase["relaxed_extjson"] as? String {
-                        expect(try Document(fromBSON: dBData).extendedJSON).to(cleanEqual(rEJ))
+            if let parseErrorTests = testFile.parseErrors {
+                for test in parseErrorTests {
+                    guard !shouldSkip(testFile.description, test.description) else {
+                        continue
+                    }
+                    let description = "\(testFile.description)-\(test.description)"
+
+                    switch BSONType(rawValue: UInt32(testFile.bsonType.dropFirst(2), radix: 16)!)! {
+                    case .invalid: // "top level document" uses 0x00 for the bson type
+                        expect(try Document(fromJSON: test.string))
+                            .to(throwError(), description: description)
+                    case .decimal128:
+                        expect(Decimal128(test.string))
+                            .to(beNil(), description: description)
+                    default:
+                        throw RuntimeError.internalError(
+                            message: "\(description): parse error tests not implemented"
+                                + "for bson type \(testFile.bsonType)"
+                        )
                     }
                 }
+            }
 
-                // for dEJ input (if it exists):
-                if let dEJ = validCase["degenerate_extjson"] as? String {
-                    // native_to_canonical_extended_json( json_to_native(dEJ) ) = cEJ
-                    expect(try Document(fromJSON: dEJ).canonicalExtendedJSON).to(cleanEqual(cEJ))
+            if let decodeErrors = testFile.decodeErrors {
+                // We only currently support initializing documents from BSON data.
+                guard testFile.description == "Top-level document validity" else {
```

Comment:
are the other cases checking e.g. just giving the portion of BSON corresponding to a single value?